xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go

```diff
@@ -414,3 +494,946 @@ func (s) TestEDSResourceRemoved(t *testing.T) {
+	defer cleanup()
+
+	// Start two test backends and extract their host and port. The first
+	// backend belongs to EDS cluster "cluster-1", while the second backend
+	// belongs to EDS cluster "cluster-2".
+	servers, cleanup2 := startTestServiceBackends(t, 2)
+	defer cleanup2()
+	addrs, ports := backendAddressesAndPorts(t, servers)
+
+	// Configure an aggregate cluster, two EDS clusters and the corresponding
+	// endpoints resources in the management server.
+	const clusterName1 = clusterName + "cluster-1"
+	const clusterName2 = clusterName + "cluster-2"
+	resources := e2e.UpdateOptions{
+		NodeID: nodeID,
+		Clusters: []*v3clusterpb.Cluster{
+			makeAggregateClusterResource(clusterName, []string{clusterName1, clusterName2}),
+			e2e.DefaultCluster(clusterName1, "", e2e.SecurityLevelNone),
+			e2e.DefaultCluster(clusterName2, "", e2e.SecurityLevelNone),
+		},
+		Endpoints: []*v3endpointpb.ClusterLoadAssignment{
+			e2e.DefaultEndpoint(clusterName1, "localhost", []uint32{uint32(ports[0])}),
+			e2e.DefaultEndpoint(clusterName2, "localhost", []uint32{uint32(ports[1])}),
+		},
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create an xDS xdsClient for use by the cluster_resolver LB policy.
+	xdsClient, close, err := xdsclient.NewWithBootstrapContentsForTesting(bootstrapContents)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	defer close()
+
+	// Create a manual resolver and push a service config specifying the use of
+	// the cds LB policy as the top-level LB policy, and a corresponding config
+	// with a single cluster.
+	r := manual.NewBuilderWithScheme("whatever")
+	jsonSC := fmt.Sprintf(`{
+			"loadBalancingConfig":[{
+				"cds_experimental":{
+					"cluster": "%s"
+				}
+			}]
+		}`, clusterName)
+	scpr := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(jsonSC)
+	r.InitialState(xdsclient.SetClient(resolver.State{ServiceConfig: scpr}, xdsClient))
+
+	// Create a ClientConn and make a successful RPC.
+	cc, err := grpc.Dial(r.Scheme()+":///test.service", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(r))
+	if err != nil {
+		t.Fatalf("Failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	// Make an RPC and ensure that it gets routed to cluster-1 since it is
+	// implicitly higher priority than cluster-2.
+	client := testgrpc.NewTestServiceClient(cc)
+	peer := &peer.Peer{}
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(peer), grpc.WaitForReady(true)); err != nil {
+		t.Fatalf("EmptyCall() RPC failed: %v", err)
+	}
+	if peer.Addr.String() != addrs[0].Addr {
+		t.Fatalf("EmptyCall() RPC routed to backend %q, want %q", peer.Addr, addrs[0].Addr)
```

Comment:
Do we want to Log Addr.String() here instead of Addr object with network etc. Would that make it more readable? Here and elsewhere.

Reply:
Ah, ok. I changed that String() method actually as a part of wrr_locality (to debug cluster resolver test breakages).